### PR TITLE
Ability to introspect self token

### PR DESCRIPTION
**Pull request by dadgar**
Ability to lookup self token, list policies and get policies attached to own token

### CODE REVIEW

File: nomad/acl_endpoint.go [hunk 1]
```diff
@@ -121,12 +121,36 @@ func (a *ACL) ListPolicies(args *structs.ACLPolicyListRequest, reply *structs.AC
 	defer metrics.MeasureSince([]string{"nomad", "acl", "list_policies"}, time.Now())
 
 	// Check management level permissions
-	if acl, err := a.srv.ResolveToken(args.AuthToken); err != nil {
+	acl, err := a.srv.ResolveToken(args.AuthToken)
+	if err != nil {
 		return err
-	} else if acl == nil || !acl.IsManagement() {
+	} else if acl == nil {
 		return structs.ErrPermissionDenied
```
Comment (schmichael) [inline nomad/acl_endpoint.go]: Doesn't `acl == nil` mean "ACLs not enabled"?

Not a big deal; an error is an error. Just seeing if there's a more precise one to use.

File: nomad/acl_endpoint.go [hunk 2]
```diff
@@ -121,12 +121,36 @@ func (a *ACL) ListPolicies(args *structs.ACLPolicyListRequest, reply *structs.AC
...
+	// If it is not a management token determine the policies that may be listed
+	mgt := acl.IsManagement()
+	var policies map[string]struct{}
+	if !mgt {
+		snap, err := a.srv.fsm.State().Snapshot()
+		if err != nil {
+			return err
+		}
+
+		token, err := snap.ACLTokenBySecretID(nil, args.AuthToken)
```
Comment (schmichael) [inline nomad/acl_endpoint.go]: Isn't this what `acl` above is?

File: nomad/acl_endpoint.go [hunk 3]
```diff
@@ -183,12 +209,42 @@ func (a *ACL) GetPolicy(args *structs.ACLPolicySpecificRequest, reply *structs.S
 	defer metrics.MeasureSince([]string{"nomad", "acl", "get_policy"}, time.Now())
 
 	// Check management level permissions
-	if acl, err := a.srv.ResolveToken(args.AuthToken); err != nil {
+	acl, err := a.srv.ResolveToken(args.AuthToken)
+	if err != nil {
 		return err
-	} else if acl == nil || !acl.IsManagement() {
+	} else if acl == nil {
 		return structs.ErrPermissionDenied
```
Comment (schmichael) [inline nomad/acl_endpoint.go]: Same question as above: is there a better error to return?

File: nomad/acl_endpoint.go [hunk 4]
```diff
@@ -183,12 +209,42 @@ func (a *ACL) GetPolicy(args *structs.ACLPolicySpecificRequest, reply *structs.S
...
 
+	// If it is not a management token determine if it can get this policy
+	mgt := acl.IsManagement()
+	if !mgt {
+		snap, err := a.srv.fsm.State().Snapshot()
+		if err != nil {
+			return err
+		}
+
+		token, err := snap.ACLTokenBySecretID(nil, args.AuthToken)
```
Comment (schmichael) [inline nomad/acl_endpoint.go]: Same q as above: how is this different from acl?

File: nomad/acl_endpoint.go (hunk 1 above)
Comment (dadgar) [reply nomad/acl_endpoint.go]: There is a check above that checks if acl's are enabled and returns a better error. So this is actually an error that will never get hit.

File: nomad/acl_endpoint.go (hunk 2 above)
Comment (dadgar) [reply nomad/acl_endpoint.go]: Haha you would think! That is the compiled ACL object and this is the raw one. The compiled one doesn't tell you the set of policies. Just the capabilities.

**Pull request merged**